16_permutations/solution/permutations-solution.spec.js

```diff
@@ -0,0 +1,82 @@
+const permutations = require('./permutations-solution');
+
+describe('permutations', () => {
+  function outcome(input, expected) {
+    const actual = permutations(input);
+
+    // Convert both arrays to strings to compare them later, excluding the order the arrays elements are in
+
+    const sterilise = (input) =>
+      input
+        .map((el) => el.toString())
+        .toSorted()
+        .toString();
+
+    return [sterilise(actual), sterilise(expected)];
+  }
+
+  let actual, expected;
+
+  afterEach(() => {
+    expect(actual).toBe(expected);
+  });
+
+  test('Works for array of size one', () => {
+    [actual, expected] = outcome([1], [1]);
+  });
+  test('Works for array of size two', () => {
+    [actual, expected] = outcome(
+      [1, 2],
+      [
+        [1, 2],
+        [2, 1],
+      ],
+    );
+  });
+  test('Works for array of size three', () => {
+    [actual, expected] = outcome(
+      [1, 2, 3],
+      [
+        [1, 2, 3],
+        [1, 3, 2],
+        [2, 1, 3],
+        [2, 3, 1],
+        [3, 1, 2],
+        [3, 2, 1],
+      ],
+    );
+  });
+  test('Works for array of size four', () => {
+    [actual, expected] = outcome(
+      [1, 2, 3, 4],
+      [
+        [
```

Comment:
You've got an extra nested array here that's not needed, but also your `outcome` function results in the test passing despite this being an incorrect expected outcome.

That being said, even after fixing this, you don't need an `outcome` helper function with an `afterEach` that uses `.toBe`. You can just use the [`.toEqual`](https://jestjs.io/docs/expect#toequalvalue) matcher that's designed exactly for this purpose.

```javascript
test('Works for array of size two', () => {
  expect(permutations([1, 2])).toEqual([
    [1, 2],
    [2, 1],
  ]);
});
```
Having all the tests as per the above format will all pass (and will also fail if you keep the unintended extra nested array in the last test "expected" array.

Reply:
> You've got an extra nested array here in the "expected array" that's not needed, but also your `outcome` function results in the test passing despite this being an incorrect expected outcome.
> 
> That being said, even after fixing this, you don't need an `outcome` helper function with an `afterEach` that uses `.toBe`. You can just use the [`.toEqual`](https://jestjs.io/docs/expect#toequalvalue) matcher that's designed exactly for this purpose.
> 
> ```js
> test('Works for array of size two', () => {
>   expect(permutations([1, 2])).toEqual([
>     [1, 2],
>     [2, 1],
>   ]);
> });
> ```
> 
> Having all the tests as per the above format will all pass (and will also fail if you keep the unintended extra nested array in the last test "expected" array.

Yes but the problem with that is, what happens if the learner's solution returns the correct array but in a different order?

> You've got an extra nested array here in the "expected array" that's not needed, but also your `outcome` function results in the test passing despite this being an incorrect expected outcome.
> 
> That being said, even after fixing this, you don't need an `outcome` helper function with an `afterEach` that uses `.toBe`. You can just use the [`.toEqual`](https://jestjs.io/docs/expect#toequalvalue) matcher that's designed exactly for this purpose.
> 
> ```js
> test('Works for array of size two', () => {
>   expect(permutations([1, 2])).toEqual([
>     [1, 2],
>     [2, 1],
>   ]);
> });
> ```
> 
> Having all the tests as per the above format will all pass (and will also fail if you keep the unintended extra nested array in the last test "expected" array.

OHhhhhhhh wow, I had no idea the following test would pass:
```javascript
    expect([1, 2]).toEqual([2, 1])
```

Reply:
Fair point I missed about the `toEqual` order. That can be remedied by sorting both sides
```javascript
test('Works for array of size two', () => {
  expect(permutations([1, 2]).sort()).toEqual(
    [
      [1, 2],
      [2, 1],
    ].sort()
  );
});
```
Note the `.sort()` for the expect is after the permutations call, not on the array passed into the permutations call.
Do that, and the order of the matched array won't matter.

Reply:
Okay, I added this